### DISCUSSION TITLE
... is declared in another module and needs to be imported"

### DISCUSSION THREAD

**Discussion by furesoft** [jbevain/cecil#868] [Q&A]
Hi,

I am writing a compiler with mono.cecil and for typeresolving I have a cache pointed to the typedefinitions. But when I want to write the assembly to a file I get an error: `"Member 'System.String' is declared in another module and needs to be imported"` for every type I reference. But this erro only occurs if I want to emit a call, in structs, parameters, variables, it works fine.

**Comment by ltrzesniewski** [comment]
Hi,

First, let's get that out of the way: I suppose you've read the [importing doc](https://github.com/jbevain/cecil/wiki/Importing)?

Here are some possible issues off the top of my head:

- Is your cache lifetime tied to the emitted module, or is it reused across several modules? Importing a type adds needed info to the metadata (most notably its assembly reference), which means you need to import each type to every module separately.

- Do you use/cache the type *returned* by `ImportReference`? If you use the `TypeReference` that you *pass* to `ImportReference`, that won't work.

- You said your cache points to type definitions, is that a mistake? You don't have the *definitions* of external types, only *references* to them.
